From fdcde4138e90f3838a93b7a7122bff1bf0f14519 Mon Sep 17 00:00:00 2001
From: "Thulasimani,Sivakumar" <sivakumar.thulasimani@intel.com>
Date: Thu, 31 Mar 2016 00:15:34 +0530
Subject: [PATCH 11/19] FOR_UPSTREAM [VPG]: drm/i915: Always perform link
 training when requested

Short pulse can be generated for automated tests requesting
link training again, such scenarios we have to perform the
request irrespective of the current status of link. This
patch adds support for this request in DP encoder.

Change-Id: I151ddf2c3ae80b66c4905eb7760435b61fff9f7c
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-30099
Signed-off-by: Sivakumar Thulasimani <sivakumar.thulasimani@intel.com>
Reviewed-on: https://android.intel.com:443/498186
---
 drivers/gpu/drm/i915/intel_dp.c | 26 +++++++++++++++-----------
 1 file changed, 15 insertions(+), 11 deletions(-)

diff --git a/drivers/gpu/drm/i915/intel_dp.c b/drivers/gpu/drm/i915/intel_dp.c
index bfadc22..1235635 100644
--- a/drivers/gpu/drm/i915/intel_dp.c
+++ b/drivers/gpu/drm/i915/intel_dp.c
@@ -4515,6 +4515,7 @@ intel_dp_check_link_status(struct intel_dp *intel_dp, bool *perform_full_detect)
 	u8 old_sink_count = intel_dp->sink_count;
 	u8 old_lane_count = intel_dp->dpcd[DP_MAX_LANE_COUNT];
 	bool ret;
+	bool check_link = false;
 	uint8_t counter = MAX_SHORT_PULSE_RETRY_COUNT;
 
 	*perform_full_detect = false;
@@ -4546,13 +4547,6 @@ intel_dp_check_link_status(struct intel_dp *intel_dp, bool *perform_full_detect)
 		return;
 	}
 
-	/* FIXME: This access isn't protected by any locks. */
-	if (!intel_encoder->connectors_active)
-		return;
-
-	if (WARN_ON(!intel_encoder->base.crtc))
-		return;
-
 	/* Try to read the source of the interrupt */
 	if (intel_dp->dpcd[DP_DPCD_REV] >= 0x11 &&
 	    intel_dp_get_sink_irq(intel_dp, &sink_irq_vector)) {
@@ -4590,12 +4584,22 @@ intel_dp_check_link_status(struct intel_dp *intel_dp, bool *perform_full_detect)
 		}
 	}
 
-	/* if link training is requested we should perform it always */
-	if ((intel_dp->compliance_test_type == DP_TEST_LINK_TRAINING) ||
-	    (!drm_dp_channel_eq_ok(link_status, intel_dp->lane_count))) {
+	/* FIXME: This access isn't protected by any locks.
+	 * if link training is requested we should perform it always
+	 */
+	if (intel_dp->compliance_test_type == DP_TEST_LINK_TRAINING) {
+		DRM_DEBUG_KMS("%s: Link training requested, retraining\n",
+				intel_encoder->base.name);
+		check_link = true;
+	} else  if (((intel_encoder->connectors_active) &&
+				(intel_encoder->base.crtc)) &&
+		(!drm_dp_channel_eq_ok(link_status, intel_dp->lane_count))) {
 		DRM_DEBUG_KMS("%s: channel EQ not ok, retraining\n",
-			      intel_encoder->base.name);
+				intel_encoder->base.name);
+		check_link = true;
+	}
 
+	if (check_link) {
 		if (IS_CHERRYVIEW(dev)) {
 			intel_dp_update_simulate_detach_info(intel_dp);
 			*perform_full_detect = true;
-- 
1.9.1

